src/Hazelcast.Net/CP/FencedLock.cs

```diff
@@ -0,0 +1,394 @@
+                    {
+                        _subsystemSession.InvalidateSession(CPGroupId, sessionId);
+                        VerifyNoLockedSessionExist(threadId);
+
+                        long duration = Clock.Milliseconds - start;
+
+                        if (duration <= 0)
+                            return InvalidFence;
+
+                    }
+                    else if (e is RemoteException { Error: RemoteError.DistributedObjectDestroyed })
+                    {
+                        throw new DistributedObjectDestroyedException(e);
+                    }
+                    else if (e is RemoteException { Error: RemoteError.WaitKeyCancelledException })
+                    {
+                        _subsystemSession.ReleaseSession(CPGroupId, sessionId);
+                        return InvalidFence;
+                    }
+                }
+                catch
+                {
+                    _subsystemSession.ReleaseSession(CPGroupId, sessionId);
+                    throw;
+                }
+            }
+        }
+
+        /// <inheritdoc/> 
+        public async Task<bool> TryLockAsync(TimeSpan timeout)
+        {
+            var fence = await TryLockAndGetFenceAsync(timeout).CfAwait();
+            return fence != InvalidFence;
+        }
+
+        /// <inheritdoc/> 
+        public async Task<bool> TryLockAsync()
+        {
+            var fence = await TryLockAndGetFenceAsync(TimeSpan.FromMilliseconds(0)).CfAwait();
+            return fence != InvalidFence;
+        }
+
+        /// <inheritdoc/> 
+        public async Task UnlockAsync()
+        {
+            var threadId = ContextId;
+            var sessionId = _subsystemSession.GetSessionId(CPGroupId);
+
+            VerifyNoLockOnThread(threadId, sessionId, false);
+
+            if (sessionId == CPSubsystemSession.NoSessionId)
+            {
+                _lockedThreadToSession.TryRemove(threadId, out var _);
+                throw new SynchronizationLockException();
+            }
+
+            try
+            {
+                Guid invocationId = Guid.NewGuid();
+                bool stillLockedByCurrentThread = await RequestUnlockAsync(sessionId, threadId, invocationId).CfAwait();
+
+                if (stillLockedByCurrentThread)
+                    _lockedThreadToSession[threadId] = sessionId;
+                else
+                    _lockedThreadToSession.TryRemove(threadId, out var _);
+
+                _subsystemSession.ReleaseSession(CPGroupId, sessionId);
+            }
+            catch (RemoteException e)
+            {
+                if (e is RemoteException { Error: RemoteError.SessionExpiredException })
+                {
+                    _subsystemSession.InvalidateSession(CPGroupId, sessionId);
+                    _lockedThreadToSession.TryRemove(threadId, out var _);
+                    throw new SessionExpiredException(e);
```

Comment:
different exception from java: throw newLockOwnershipLostException(sessionId);

Reply:
As current behavior, throws only RemoteException. Fixed at https://github.com/hazelcast/hazelcast-csharp-client/commit/993cabbcb9bac708cc38ea83239b98cf19ead875

Reply:
Will be discussed in another issue.

Reply:
Related issue https://github.com/hazelcast/hazelcast-csharp-client/issues/592